### ISSUE TITLE
NoReturn annotation on method not honored

### ISSUE THREAD

**Issue by Jasha10** [microsoft/pyright#3001]
**Describe the bug**
When a class has a method annotated as `NoReturn`, calling that method in client code does not respect the `NoReturn` annotation.

**To Reproduce**
```python
# example.py
from time import sleep
from typing import NoReturn

class MyClass:
    def no_return(
        self,
    ) -> NoReturn:
        while True:
            sleep(1)

def client_code() -> NoReturn:
    instance = MyClass()
    instance.no_return()
```
```text
$ pyright example.py
...
Found 1 source file
/.../example.py
  /.../example.py:14:22 - error: Function with declared type of "NoReturn" cannot return "None" (reportGeneralTypeIssues)
1 error, 0 warnings, 0 informations
```

**Expected behavior**
There should be no error above. The `instance.no_return` method call should be recognized as NoReturn, meaning that `client_code` is also typed as `NoReturn`.

**Additional context**
pyright version 1.1.218

**Comment by erictraut** [comment]
This is a tricky case to handle given the internal architecture of pyright. It builds a code flow graph to answer questions like "is it possible to reach the end of this function block without returning or raising an exception"? The code flow graph is also used to perform type inference and local type narrowing. In your example, `instance` is an unannotated local variable, so its type needs to be inferred, but type inference relies on the code flow graph, the code flow graph depends on whether `instance.no_return()` returns a `NoReturn` type, and the type of `instance.no_return()` depends on the inferred type of `instance` . This circularity is difficult to resolve in the general case, so pyright uses some special-case handling to attempt to handle common cases. This special-case logic wasn't able to handle the case in your code example.

I've found a way to handle this particular case, and it will be included in the next release.

In the meantime, you can work around the false positive error by adding a type annotation for the `instance` variable. This breaks the circularity because type inference is no longer required.

```python
instance: MyClass = MyClass()
```

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.219, which I just published. It will also be in the next release of pylance.

**Comment by erictraut** [comment]
The fix for this bug caused a significant performance regression and some crashes, so I need to back it out and look for other possible solutions.

**Comment by q0w** [comment]
Also casting to `NoReturn` does not work
```python
from typing import cast, NoReturn, Callable

def get_d() -> dict | None:
    pass

def t():
    pass

x = get_d() # return dict or None

if not x:
  cast(Callable[[], NoReturn], t)()

a = x.get('b') # error: "get" is not a known member of "None"
```

**Comment by erictraut** [comment]
This will be addressed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.243, which I just published. It will also be included in the next release of pylance.
